README
NAME
    Types::LoadableClass - type constraints with coercion to load the class

SYNOPSIS
       package MyClass;
       use Moose;  # or Mouse, or Moo, or whatever
       use Types::LoadableClass qw/ LoadableClass /;
   
       has foobar_class => (
          is       => 'ro',
          required => 1,
          isa      => LoadableClass,
       );
   
       MyClass->new(foobar_class => 'FooBar'); # FooBar.pm is loaded or an
                                               # exception is thrown.

DESCRIPTION
    A Type::Tiny-based clone of MooseX::Types::LoadableClass.

    This is to save yourself having to do this repeatedly...

      my $tc = subtype as ClassName;
      coerce $tc, from Str, via { Class::Load::load_class($_); $_ };

    Despite the abstract for this module, `LoadableClass` doesn't actually
    have a coercion, so no need to use `coerce => 1` on the attribute. Rather,
    the class gets loaded as a side-effect of checking that it's loadable.

  Type Constraints
    `ModuleName`
        A subtype of `Str` (see Types::Standard) representing a string that is
        a valid Perl package name (according to Module::Runtime).

    `LoadableClass`
        A subtype of `ModuleName` that names a module which is either already
        loaded (according to Class::Load), or can be loaded (by Class::Load).

    `LoadableRole`
        A subtype of `LoadableClass` that names a module which appears to be a
        role rather than a class.

        (Because this type constraint is designed to work with Moose, Mouse,
        Moo, or none of the above, it can't rely on the features of any
        particular implementation of roles. Therefore is needs to use a
        heuristic to detect whether a loaded package represents a role or not.
        Curently this heuristic is the absence of a method named `new`.)

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Types-LoadableClass>.

SEE ALSO
    Types::Standard, MooseX::Types::LoadableClass, Class::Load,
    Module::Runtime.

AUTHOR
    Dagfinn Ilmari Mannsåker <ilmari@ilmari.org>.

    Improvements and packaging by Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Dagfinn Ilmari Mannsåker, Toby
    Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Types-LoadableClass
===================

Created:      2013-10-03
Home page:    <https://metacpan.org/release/Types-LoadableClass>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Types-LoadableClass>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.002	2013-10-03

 [ Packaging ]
 - List correct dependencies.

0.001	2013-10-03	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "type constraints with coercion to load the class",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.013, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Types-LoadableClass",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       requires => {
                                         "Class::Load"     => 0,
                                         "Module::Runtime" => 0,
                                         "perl"            => 5.006001,
                                         "Type::Library"   => 0.026,
                                         "Type::Utils"     => 0.026,
                                         "Types::Standard" => 0.026,
                                       },
                                     },
                        test      => {
                                       recommends => { Moo => "1.000000" },
                                       requires   => { "Test::Fatal" => 0, "Test::More" => 0.96, "Test::Requires" => 0.06 },
                                       suggests   => { Moose => "2.0000" },
                                     },
                      },
  "provides"       => {
                        "Types::LoadableClass" => { file => "lib/Types/LoadableClass.pm", version => 0.002 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=Types-LoadableClass",
                                        },
                        homepage     => "https://metacpan.org/release/Types-LoadableClass",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          web  => "https://github.com/tobyink/p5-types-loadableclass",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/Types-LoadableClass/project",
                      },
  "version"        => 0.002,
  "x_contributors" => ["Dagfinn Ilmari Manns\xE5ker (ILMARI) <ilmari\@cpan.org>"],
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

